### PR TITLE
ENH: Relative tolerance for GLM convergence

### PR DESCRIPTION
**Pull request by thequackdaddy**
Hello, 

I'd like to propose changing how you do convergence checking for GLM IRLS fits. 

A few notes:

1. Currently, I think the iteration check for IRLS has an off-by-1 error. The 1st run check's the 0th iteration vs. the -1st iteration. The 2nd  iteration will check the 1st and 0th. The 3rd with check the 1st and the 2nd, and so on. 
2.  I've been playing with statsmodels on some insurance datasets. I typically get really large datasets (hundreds of thousands of observations). I'm having difficulty getting convergence in reasonable times because the convergence tolerance is an absolute `1e-8`. I think this makes more sense because deviance increases as you add more observations. On a large dataset, waiting for deviance changes as small as 1e-8 requires many (essentially useless) iterations. 
3.  As a result of 2., a few of the tests break. I think my tolerance is a little more lenient so you lose a few significant digits. I'll work through that, but I want to know if this is a PR that you'd consider before I invest the time in fixing that. 

Thanks for your consideration.

### CODE REVIEW

**Comment by josef-pkt** [comment]
Thanks for pointing this out

some unsorted thoughts

- in general I don't like replacing absolute tol with relative tol. I think we could add rtol similar to np.allclose
- using only rtol for deviance might run into problems when deviance is zero or goes to zero. Deviance is zero in saturated log-linear model for example.
- I didn't realize that the convergence criterion is in terms of deviance and not parameters. I thought we usually check convergence in parameters. (Problem in the case of complete separation is that deviance or loglike can converge but not params).
- deviance scales with nobs, so maybe there the deviance criterion should divide by nobs (average per unit)

a tip

- (requires master but not recent, IIRC):  Did you try fit(... method='bfgs') or method='newton'.
  We have no the main scipy optimizers also available in GLM. In some cases it converges faster.

scipy optimizers all have tolerance absolute and in terms of parameters, IIRC

as a workaround: 
does it help in your case if you set `tol=1e-8 * nobs`. I guess it might need another factor for full scaling with nobs.

We need to do something, but I'm not quite sure what.

(I haven't looked at off-by-1 yet.)

**Comment by josef-pkt** [comment]
@kshedden  Any opinion on this?

File: statsmodels/genmod/generalized_linear_model.py [hunk 1]
```diff
@@ -41,7 +41,13 @@
 
 
 def _check_convergence(criterion, iteration, tol):
-    return not (np.fabs(criterion[iteration] - criterion[iteration-1]) > tol)
+    if iteration == 0:
+        return False
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: Why did you need to special case iteration 0?
We could have converged after one iteration.

**Comment by josef-pkt** [comment]
GLM is a core model and it has been this way for a long time, most likely since the beginning.
So we want to switch only once and not experiment. We should have a general policy for this.

File: statsmodels/genmod/generalized_linear_model.py (hunk 1 above)
Comment (thequackdaddy) [inline statsmodels/genmod/generalized_linear_model.py]: Maybe I'm confused, but the 0th element of `criterion[iteration]` is set to be `np.inf`. The first iteration will compare `np.inf` vs. the deviance after the first IRLS iteration. So, it shouldn't converge ever. 

So I think you are right, we don't need the `iteration == 0` case as `np.inf - x` should always be > tol.

No?

**Comment by josef-pkt** [comment]
RLM is similar and the convergence criterion is on the value of the objective function, but it has an option for choosing it `conv='dev'`.
(RLM was largely modeled the same way as GLM)

File: statsmodels/genmod/generalized_linear_model.py (hunk 1 above)
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: I think you are right in general. We need one extra iteration to check convergence even with gaussian and OLS, because we don't know yet that we already convert. (I had forgotten about this.)

However, following your off-by-1 argument, I think the initial np.inf is redundant, since we initialized with the starting estimate. If the starting values (e.g. user provided start_params) are already optimal, then we only need the iteration=0 round to verify that we are converged.

**Comment by josef-pkt** [comment]
I added prio-high label for some solution to the convergence criterion.
Absolute tolerance for deviance or objective function doesn't make sense if the objective is proportional to nobs.

**Comment by josef-pkt** [comment]
I opened an issue #2825 for the record
According to julia GLM issues, R uses `abs(dev - devold)/(abs(dev) + 0.1) < convTol`
which looks like allclose with fixed ratio of atol / rtol

**Comment by thequackdaddy** [comment]
Yes, that's the impetus for my original pull request. I was under the
impression the +0.1 was just to avoid a divide by zero error, but I could
see how it acts as a fixed atol as well.

On Thursday, February 18, 2016, Josef Perktold <notifications@github.com>
wrote:

> I opened an issue #2825
> <https://github.com/statsmodels/statsmodels/issues/2825> for the record
> According to julia GLM issues, R uses abs(dev - devold)/(abs(dev) + 0.1)
> < convTol
> which looks like allclose with fixed ratio of atol / rtol
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/statsmodels/statsmodels/pull/2803#issuecomment-186007062>
> .
>


-- 
Peter Quackenbush
pquack@gmail.com

**Comment by thequackdaddy** [comment]
As another aside, when I was fitting GLM's with R, I would get convergence in 5-10 iterations. statsmodels was taking 25 or so for the same data.

**Comment by josef-pkt** [comment]
About the number of iterations it takes:
I think an improvement can only come from better starting values. R uses QR and we use SVD by default but I don't think that should make a difference for convergence. The algorithm should be the same standard IRLS.
But it will depend a lot on the convergence criterion.

Do you know what the speed difference is between statsmodels and R?

**Comment by josef-pkt** [comment]
I think this PR looks good now. I think we can go the direction of RLM and add **another** option to choose what is used for convergence, e.g. so users or we, if or when we want to change the default, can switch to convergence in params.

I think convergence in params, and score, are a lot more important than convergence in deviance, sum of squares or loglike.

**Comment by josef-pkt** [comment]
I haven't looked at the test results yet, and didn't check what the default tolerances should be.

**Comment by thequackdaddy** [comment]
Ok. I'm going to go radio silent on this PR then and leave it up to your discretion how you want to implement. I think the failing tests are due to...

1. Before we had an off-by-one error. (Actually, I'm thinking it was really 2 off-by-one errors, so off-by-2) This caused IRLS iterations to continue beyond when they should have stopped because of the convergence criterion
2.  Log - Binomial is unstable in the gradient vs. IRLS test

If you wish for me to add stuff let me know. 

I hadn't thought about using other metrics for the tolerance calculation. Off the top of my head, parameters, log-likelihood, and scores sound simple enough. But I'd want a list before I invest any time programming it.

**Comment by thequackdaddy** [comment]
One other thought... log-likelihood seems especially compelling considering the basis for the calculus underlying this is MLE. I'd be interested if we could find a use case where an iteration causes likelihood to fall. If MLE falls, that might be enough to make us want to break the algorithm as the fit is probably garbage.

**Comment by josef-pkt** [comment]
loglikelihood and deviance are essentially the same, up to constant factors.

The only two options right now would be `deviance` or `params`, loglikelihood (`llf`) could be added for consistency but will be essentially the same as deviance in the GLM case.

Our IRLS is plain vanilla and doesn't do backtracking nor checking for no improvement.
Julia again: They have one issue where deviance goes down, most likely because floating point precision or rounding problems, and then the backtracking fails.
I haven't heard of many problems with our IRLS, and since we have the scipy optimizers now as alternative, with some robust slow, or less robust fancy optimizers, I didn't consider improving our IRLS as high priority.

There are some substantial differences between IRLS and scipy optimizers for non-canonical links. In that case the Hessian and the Fisher information matrix are not the same and they don't solve exactly the same problem. (I haven't investigated that in details.)

**Comment by josef-pkt** [comment]
I'm starting to look a bit at the test failures

statsmodels.discrete.tests.test_constrained.TestGLMLogitConstrained2.test_basic_method
uses spector data which has only 32 observations.
difference in 5th decimal in bse, reference result are from Stata

**Comment by thequackdaddy** [comment]
I loosened some of the tests to show it now passes. Like I said, I'm fairly confident that before IRLS would continue to run well after the convergence criterion was met. That made some results too good. 

The tests seem to be passing here. The fails are the other one's we've had sitting around.

**Comment by josef-pkt** [comment]
The problem is that the old setting affects convergence very differently depending on the deviance, and indirectly the sample size. Even though we loosen for large sample sizes, we shouldn't loosen (too much) for small sample sizes.

I still don't have a good overview how this affects different cases. The increase in the tolerance in here might be too large. Furthermore, I think I'd like to move to using params as default convergence target.

There are two things in this PR: 
- adding atol and rtol as separate keywords, which could be added/merged in a backwards compatible way
- changing the default tolerance. This impacts what numbers we get back, as in the unit tests, and I think requires more investigation into different criteria and how they affect various cases.

One extra check to verify how good the convergence is, is to look at the resulting `score`

**Comment by thequackdaddy** [comment]
Ok. Well, I'll leave this into your hands. You have a better understanding on the fancy-shmancy math anyways. 

Let me know if you want me to try something. I'm happy to experiment with using param convergence as the standard, but a little concerned that will break even more tests. If you don't mind me taking a whirl at it, let me know. In that event I'd probably revert the last change I made and then change the `_check_convrgence` function to incorporate the other values available in `fit_history`.

**Comment by josef-pkt** [comment]
I think the easiest way forward is to add the option that lets user choose the convergence target, i.e. add `params`. And then experiment in a script or notebook with different cases.

This PR and issue have convinced me that we need to change the default criterion, which will always have the affect on some numbers. However, I would prefer that we have to do this only once, so I would like to understand better what we are switching to.
Given that GLM is an established core model, I would rather not experiment across releases, that's why I'm a bit more conservative in this.

**Comment by thequackdaddy** [comment]
Ok, I lied. I was able to find some time to work on this today. 

I'm probably a little better with python than I was previously, so I sort of started from scratch here. 

3 real changes behavioral changes:
* If you specify `atol` or `rtol`, those will override, for IRLS fits only, `tol`. If you specify `rtol`, relative tolerance is used. 
* If you specify `tol_criterion = 'params'`, IRLS fits will run until `params` have converged. 
* I fixed the 2 "off by 1 error" that were causing additional iterations to run after the convergence criterion had been satisfied. Note that this breaks a few tests. The solution that I ended up going with was to specify a higher `tol` for the `fit` methods in those tests to run the additional iterations to match.

As I'm reviewing this, I realized a small typo in my documentation, so I'll fix that, but I think this is ready for review. 

Thanks.

File: statsmodels/discrete/tests/test_constrained.py [hunk 2]
```diff
@@ -357,14 +357,16 @@ def setup_class(cls):
...
 
         constr = 'C(agecat)[T.4] = C(agecat)[T.5]'
         lc = patsy.DesignInfo(mod.exog_names).linear_constraint(constr)
-        cls.res1 = fit_constrained(mod, lc.coefs, lc.constants)
+
+        cls.res1 = fit_constrained(mod, lc.coefs, lc.constants,
+                                   fit_kwds={'tol': 1e-10})
         cls.constraints = lc
-        cls.res1m = mod.fit_constrained(constr)._results
+        cls.res1m = mod.fit_constrained(constr, tol=1e-10)._results
```
Comment (josef-pkt) [inline statsmodels/discrete/tests/test_constrained.py]: it's better to be explicit and use `atol` instead of `tol` in these kinds of tests

File: statsmodels/genmod/generalized_linear_model.py [hunk 3]
```diff
@@ -963,9 +985,9 @@ def _fit_irls(self, start_params=None, maxiter=100, tol=1e-8,
 
         # first guess on the deviance is assumed to be scaled by 1.
         # params are none to start, so they line up with the deviance
-        history = dict(params=[None, start_params], deviance=[np.inf, dev])
+        history = dict(params=[np.inf, start_params], deviance=[np.inf, dev])
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: do we need the `inf` because we also check the starting values for convergence?
It looks to me, that we update_history once before the first convergence checks. So the inf might be redundant in that case.

Comment (thequackdaddy) [inline statsmodels/genmod/generalized_linear_model.py]: I added `inf` because `allclose` in the `_check_convergence` function didn't like comparing `None` to an array.

Comment (thequackdaddy) [inline statsmodels/genmod/generalized_linear_model.py]: Wait now that I think about it, I think it was actually `allclose` that was in one of my tests. See this line. 

https://github.com/statsmodels/statsmodels/pull/2803/files#diff-bacf4e7dec296273aba74b04be3c42f8R1687

I presume I could just have it skip the first test... Duh. Let me fix that.

Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: maybe I wasn't clear:
does starting with the initial values work?    (with neither None nor inf)
`history = dict(params=[start_params], deviance=[dev])`

Comment (thequackdaddy) [inline statsmodels/genmod/generalized_linear_model.py]: Oh... that will likely break some tests. It will definitely break the test that I wrote. 

I can experiment with that if you like.

Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: Leave it as you have it right now. inf is more explicit, and RLM also uses inf

File: statsmodels/discrete/tests/test_constrained.py (hunk 2 above)
Comment (thequackdaddy) [inline statsmodels/discrete/tests/test_constrained.py]: Fair point. I changed this in the most recent revision.

File: statsmodels/genmod/generalized_linear_model.py [hunk 4]
```diff
@@ -877,6 +878,23 @@ def fit(self, start_params=None, maxiter=100, method='IRLS', tol=1e-8,
             values for gradient optimization.  Only relevant if
             `method` is set to something other than 'IRLS'.
 
+        IRLS Parameters
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: we cannot use nonstandard titles/headers in the docstrings because they are not recognized by the doc rendering

AFAICS either use `Other Parameters` or no headers for this section just continue with parameters
section 7 https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt
AFAIK we don't use `Other Parameters` anywhere so I'm not sure how it works.

**Comment by josef-pkt** [comment]
that's all I can see right now. I will go over it in a bit more details on Friday and merge.

One thing I haven't looked at or thought about is whether we want to do something with the `tol` parameter in the signature given that atol and rtol are hiding in the `kwds`.

File: statsmodels/genmod/generalized_linear_model.py (hunk 4 above)
Comment (thequackdaddy) [inline statsmodels/genmod/generalized_linear_model.py]: Whoops. Didn't realize that was a "thing". I presume you could use any header. Thanks for that link. I tried to follow

**Comment by thequackdaddy** [comment]
Sounds good, as always Josef, thank you for your insights and helpfulness. I really appreciate it.

File: statsmodels/genmod/generalized_linear_model.py [hunk 5]
```diff
@@ -877,6 +878,21 @@ def fit(self, start_params=None, maxiter=100, method='IRLS', tol=1e-8,
...
+            The relative tolerance criterion that must be satisfied. Defaults
+            to 0 which means ``rtol`` is not used. Convergence is attained
+            when:
+            :math:`rtol * prior + atol > abs(current - prior)`
+        tol_criterion : str, optional
+            Defaults to ``'deviance'``. Can optionally be ``'params'``.
+
         Notes
         -----
         This method does not take any extra undocumented ``kwargs``.
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: can you delete this Notes section, 
AFAICS, this is outdated because there are also optimization kwargs for fit_gradient

Comment (thequackdaddy) [inline statsmodels/genmod/generalized_linear_model.py]: I will. To avoid having a billion commits, I'll commit that in a few hours in case my OCD comes back.

**Comment by josef-pkt** [comment]
Given that there were a few back and forth commits, I think this could be squashed. I don't see good breaks in the list of commits, so it could be squashed into a single commit.

Can you do that?

To clarify and summarize, the default has not changed at all. The only change in default behavior is coming from fixing the off-by-one error. Is that correct.

I think I've seen enough, ready for merge, preferably after squashing, and when travis is as green as it gets.

**Comment by thequackdaddy** [comment]
Yes, I can squash. I'm not as useless at that as I used to be.

* Default criterion of `atol=1e-8` on `devaince` is unchanged. We are using `np.allclose` instead of a homebrewed equivalent.
* A change in results is because this will do 1 less iteration than the previous version because of the correction of the off by one error.
* You can use `tol_criterion = 'params'` to check convergence of `params` instead of `deviance`. As far as I know, every tool I've used has defaulted to `deviance` (R and 2 proprietary commercial products are the only other tools I've run GLMs on.)

**Comment by thequackdaddy** [comment]
Squashed. Feel free to merge at your leisure.

**Comment by josef-pkt** [comment]
@thequackdaddy   Thanks for another improvement to GLM
merging

**Comment by josef-pkt** [comment]
It would be useful for future reference and recommendation and possible changes in default settings if you could add some feedback comments, once you get experience with how well this works and have recommendation for good values for rtol in your kind of use cases. 

Based on the discussion here I still think we need to change the defaults, but this PR is backwards compatible so much easier to merge.

**Comment by thequackdaddy** [comment]
I don't disagree. Honestly, what I've been doing is just setting `max_iter=10`, which has worked well enough. When you have 100,000+ records, and the models I'm building are relatively clean and straightforward, I usually get convergence pretty quickly. But I like the idea of using `rtol` more, personally. 

With that said, I can understand the advantages of `atol`. I personally won't use it, but I am not the only user here...

Thanks again.

**Comment by josef-pkt** [comment]
The crucial qualifier is "relatively clean and straightforward".

Overall, I think the current (a)tol on deviance is a bad idea (even if many packages use it), if we want reliable parameter estimates, or an indication and convergence warnings if the parameters don't converge cleanly and straightforwardly.

**Comment by thequackdaddy** [comment]
Fair point. I had never considered using `params` as the convergence criterion. As I think about that more, rtol on `params` is intriguing. 

(Thinking out loud below)

The one issue that I have with this is that in practicality, I set one column in the dataset to a constant 1 (`add_constant`). The rest of the columns are indicators for presence or absence of a feature (male/female, discount, etc.) or a standardized interval value (policyholder age, standardized). When modeling something like severity per claim or average loss per year, the `param` associated with the constant column will be a large positive number (in the hundreds to ten-thousands) while the other `params` will be centered around 0 with params probably < 0.3 or so (generally speaking). So `atol` on `params` would not be ideal because it might take a long to converge the constant column because we're talking about 12-13 digits. `rtol` is a different story, but I worry about what would happen if you have an alias in the dataset (which happens). There, I've seen one of the aliased parameters numbers really close to zero (think 1e-100), so getting convergence on `rtol` would be problematic too. 

With that said, at this point I statsmodels can fit a GLM with 3.5 million records and 25 columns on a virtual machine with a eight 1.9 GHZ processors in about a 1.5 minutes when I limit to 10 iterations. After this change, I think that time is down to around about one minute (its about 5-7 iterations for `deviance` to converge... not sure about `params`.) That's about as good as any commercial package I've dealt with. The bottleneck that I still struggle with in model-constructing is the converting my MSSQL database into numpy arrays/pandas dataframes. The real issue isn't so much the data transfer but creating all the dummy variables and standardized intervals quickly. I've played with a few approaches and all have been wholly unsatisfying, but I stumbled across a new idea a few days ago that I need to play with.

**Comment by josef-pkt** [comment]
Did you try plain numpy for creating the dummy variables?
Do you know the unique values for the categorical variables in advance?
But that discussion is in a different issue.

for tolerance I think that in cases like yours, both atol and rtol would have to be set. For example in unit tests I usually fine tune both for assert_allclose. My guess is that a default of rtol=atol=1e-8 or 1e-6 would work in most cases for params. However, atol might be too large if we have a tiny param but a large effect because the scale of the exog is large. The proper tolerance might be in terms of standardized coefficients, i.e. `etol * exog.std()`, if exog varies a lot in scale. Bad scaling should be avoided in any case.
`score` could be checked with an atol because it is supposed to be zero at the optimum, and if it's not close to zero, then it also indicates some convergence problems.
